R/board_gcs.R

```diff
@@ -0,0 +1,262 @@
+#'   `predefinedAcl` and `upload_type` through the dots of `pin_write()`.
+#' * `board_gcs()` is powered by the googleCloudStorageR package, which is a
+#'   suggested dependency of pins (not required for pins in general). If
+#'   you run into errors when deploying content to a server like
+#'   <https://shinyapps.io> or [Connect](https://posit.co/products/enterprise/connect/),
+#'   add `library(googleCloudStorageR)` to your app or document for [automatic
+#'   dependency discovery](https://support.posit.co/hc/en-us/articles/229998627-Why-does-my-app-work-locally-but-not-on-my-RStudio-Connect-server).
+#'
+#' @inheritParams new_board
+#' @param bucket Bucket name. You can only write to an existing bucket. Defaults
+#'   to [googleCloudStorageR::gcs_get_global_bucket()].
+#' @param prefix Prefix within this bucket that this board will occupy.
+#'   You can use this to maintain multiple independent pin boards within
+#'   a single GCS bucket. Will typically end with `/` to take advantage of
+#'   Google Cloud Storage's directory-like handling.
+#' @export
+#' @examples
+#' \dontrun{
+#' board <- board_gcs()
+#' board %>% pin_write(mtcars)
+#' board %>% pin_read("mtcars")
+#'
+#' # A prefix allows you to have multiple independent boards in the same pin.
+#' board_sales <- board_gcs("company-pins", prefix = "sales/")
+#' board_marketing <- board_gcs("company-pins", prefix = "marketing/")
+#' # You can make the hierarchy arbitrarily deep.
+#'
+#' # Pass arguments like `predefinedAcl` through the dots of `pin_write`:
+#' board %>% pin_write(mtcars, predefinedAcl = "publicRead")
+#' }
+board_gcs <- function(bucket = googleCloudStorageR::gcs_get_global_bucket(),
+                      prefix = NULL,
+                      versioned = TRUE,
+                      cache = NULL) {
+
+  check_installed("googleCloudStorageR")
+
+  # Check that have access to the bucket
+  invisible(googleCloudStorageR::gcs_get_bucket(bucket))
+
+  cache <- cache %||% board_cache_path(paste0("gcs-", bucket))
+  new_board_v1("pins_board_gcs",
+               name = "gcs",
+               bucket = bucket,
+               prefix = prefix,
+               cache = cache,
+               versioned = versioned
+  )
+}
+
+board_gcs_test <- function(...) {
+
+  skip_if_missing_envvars(
+    tests = "board_gcs()",
+    envvars = c("PINS_GCS_PASSWORD")
+  )
+
+  rlang::check_installed("sodium")
+  path_to_encrypted_json <- fs::path_package("pins", "secret", "pins-gcs-testing.json")
+  raw <- readBin(path_to_encrypted_json, "raw", file.size(path_to_encrypted_json))
+  pw <- Sys.getenv("PINS_GCS_PASSWORD", "")
+  json <- sodium::data_decrypt(
+    bin = raw,
+    key = sodium::sha256(charToRaw(pw)),
+    nonce = secret_nonce()
+  )
+  googleCloudStorageR::gcs_auth(json_file = rawToChar(json))
+
+  board_gcs("pins-dev",
+            cache = tempfile(),
+            ...
+  )
+}
+
+#' @export
+pin_list.pins_board_gcs <- function(board, ...) {
+  NA
+}
```

Comment:
Like I discussed earlier in this PR, no `pin_list()` for the time being because we don't have good tooling for handling the flat namespace in a bucket. This is different from AWS and Azure. We can try coming back to this later if it is a high priority issue for folks.

Reply:
I guess to enable it this would need to download a list of all objects, then parse through it locally for those ending with /.  Its doable, but could get slow if a very large (10k+) number of objects

Reply:
You can see how we handle that for [AWS](https://github.com/rstudio/pins-r/blob/b50eb57ae5879e28f85e42940f8f4944b49ff471/R/board_s3.R#L146-L156) and [Azure](https://github.com/rstudio/pins-r/blob/main/R/board_azure.R#L241-L250). Both of those rely on underlying functionality already available in paws.storage and AzureStor (ways to identify directory-like structure, common prefixes, etc). I don't think I want to move forward with writing that parsing code here in pins for now; maybe we can collaborate in the future on adding features like this to googleCloudStorageR if it is a high priority for users.

Reply:
Why can't we use `gcs_list_objects()` plus a prefix? Because it returns all objects "recursively"?

Reply:
Yes, that's right:

``` r
library(pins)
googleCloudStorageR::gcs_auth("~/google-pins.json")
b <- board_gcs("pins-testing")
googleCloudStorageR::gcs_list_objects(bucket = b$bucket)
#>                                                         name      size
#> 1        big-numbers/20230110T204627Z-b89e8/big-numbers.json  46 bytes
#> 2                big-numbers/20230110T204627Z-b89e8/data.txt 187 bytes
#> 3              small-numbers/20230110T204626Z-4d1eb/data.txt 189 bytes
#> 4     small-numbers/20230110T204626Z-4d1eb/small-numbers.rds  61 bytes
#> 5              small-numbers/20230110T204628Z-c3943/data.txt 191 bytes
#> 6    small-numbers/20230110T204628Z-c3943/small-numbers.json  23 bytes
#> 7            test-kcH7P3OJjX/20230110T205041Z-6e4eb/data.txt 192 bytes
#> 8 test-kcH7P3OJjX/20230110T205041Z-6e4eb/test-kcH7P3OJjX.rds  44 bytes
#>               updated
#> 1 2023-01-10 20:46:28
#> 2 2023-01-10 20:46:28
#> 3 2023-01-10 20:46:26
#> 4 2023-01-10 20:46:27
#> 5 2023-01-10 20:46:29
#> 6 2023-01-10 20:46:29
#> 7 2023-01-10 20:50:41
#> 8 2023-01-10 20:50:41
googleCloudStorageR::gcs_list_objects(bucket = b$bucket, prefix = "big-numbers")
#>                                                  name      size
#> 1 big-numbers/20230110T204627Z-b89e8/big-numbers.json  46 bytes
#> 2         big-numbers/20230110T204627Z-b89e8/data.txt 187 bytes
#>               updated
#> 1 2023-01-10 20:46:28
#> 2 2023-01-10 20:46:28
```

<sup>Created on 2023-01-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

In the packages for AWS and Azure, there is existing support for identifying directory-like structures or to find "common prefixes" but googleCloudStorageR doesn't have that as of today.